File: testapp/src/main/java/com/saantiaguilera/featureflags/provider/java/CacheProvider.java
package com.saantiaguilera.featureflags.provider.java;

import com.saantiaguilera.featureflags.FeatureFlag;
import com.saantiaguilera.featureflags.FeatureFlagProvider;
import com.saantiaguilera.featureflags.FeatureFlagResult;
import com.saantiaguilera.featureflags.FeatureFlagResultExtension;
import com.saantiaguilera.featureflags.provider.kotlin.CacheRepository;

import org.jetbrains.annotations.NotNull;

/**
 * Java example of one of the providers. Note that it's pretty much the same besides the APIs and
 * the usage of {@link FeatureFlagResultExtension} for extension functions.
 */
public class CacheProvider implements FeatureFlagProvider {

    private final CacheRepository repository;

    public CacheProvider(final CacheRepository repository) {
        this.repository = repository;
    }

    @NotNull
    @Override
    public FeatureFlagResult provide(@NotNull final FeatureFlag feature) {
        final Boolean result = repository.getFeatures().getOrDefault(feature.getKey(), null);

        if (result == null) {
            return FeatureFlagResultExtension.createMissingResult(feature.getValue());
        } else {
            return FeatureFlagResultExtension.createExistingResult(result);
        }
    }

}

File: testapp/src/main/java/com/saantiaguilera/featureflags/provider/kotlin/CacheProvider.kt
package com.saantiaguilera.featureflags.provider.kotlin

import com.saantiaguilera.featureflags.FeatureFlag
import com.saantiaguilera.featureflags.FeatureFlagProvider
import com.saantiaguilera.featureflags.FeatureFlagResult

/**
 * Basic provider that hits a cache on every request
 */
class CacheProvider(private val repository: CacheRepository) : FeatureFlagProvider {

    override fun provide(feature: FeatureFlag): FeatureFlagResult {
        return repository.getFeatures()[feature.key]
            ?.let { FeatureFlagResult(feature, it) }
            ?: FeatureFlagResult(feature)
    }

}

/**
 * This sample uses a "cache", but it could be anything (db / api-call / etc). Bear in mind
 * we access it for each invoke in this sample.
 */
interface CacheRepository {

    fun getFeatures(): Map<String, Boolean>

}